### PR TITLE
[MRG] Refactor log likelihood.

### PR DESCRIPTION
**Pull request by jasmainak**
This is built on top of https://github.com/glm-tools/pyglmnet/pull/216.

@pavanramkumar there are some differences between the two functions that I don't understand. Perhaps you could explain to me over chat

### CODE REVIEW

**Comment by pavanramkumar** [comment]
@jasmainak the `_logL()` within `pyglmnet.py` is called by `_loss()` which is only used to track the loss and determine convergence. the `_logL()` within `utils.py` is used for scoring the model.

In the former case, it makes sense to think of the log likelihood as a function of the paremeters (`beta0`, `beta`) since that is what you are optimizing wrt to. In the latter case, we want to score the prediction against the test data. Here, you want to separate the scoring from the estimation as much as possible. There are also numerical issues such as adding an `eps` parameter to make sure we don't get -`Inf` or `NaN`. 

Hence, the difference in API of the two functions. Let me know your suggestions to DRY them.

**Comment by jasmainak** [comment]
hmm ... I was thinking of abstracting away the likelihood as something that is a property of a distribution. The loss function that we have of course takes the parameters `beta` as that is how we parametrize it. 

I'm also thinking in terms of trivializing how future contributors can add a new distribution. It should be as simple as updating the log likelihood and it's derivative. The link function and rest should all be taken care of automatically because of our API design.

Worst case, we can add a parameter (`mode='scoring'` for example) that chooses the code path depending on weather we are scoring or doing the loss function.

File: pyglmnet/utils.py [hunk 1]
```diff
@@ -4,40 +4,13 @@
...
 from copy import copy
-from scipy.stats import norm
 import logging
 
 
 logger = logging.getLogger('pyglmnet')
 logger.addHandler(logging.StreamHandler())
 
 
-def probit(z):
```
Comment (jasmainak) [inline pyglmnet/utils.py]: @pavanramkumar you will notice this trend that I like deleting lines a lot ;)

Quoting zen of python: "Flat is better than nested." We don't really need these wrappers and in any case I don't see why they had to be public ...

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -14,10 +15,11 @@ def _lmb(distr, beta0, beta, X, eta):
     return _qu(distr, z, eta)
 
 
-def _qu(distr, z, eta):
+def _qu(distr, z, eta, return_grad=False):
     """The non-linearity."""
-    if distr == 'softplus':
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: 'softplus' is not really a distribution, no? Maybe we can say `'poisson-softplus'` and `'poisson-linearized'` to be more explicit

File: doc/index.rst [hunk 3]
```diff
@@ -3,9 +3,9 @@
    You can adapt this file completely to your liking, but it should at least
    contain the root `toctree` directive.
 
-==========================================================================
-Python implementation of elastic-net regularized generalized linear models
```
Comment (jasmainak) [inline doc/index.rst]: @pavanramkumar I removed elastic-net from the title. I think we have no expanded beyond just elastic nets. If you're not okay with it, I can put it back.

File: pyglmnet/pyglmnet.py [hunk 4]
```diff
@@ -26,40 +28,39 @@ def _qu(distr, z, eta):
...
     elif distr == 'binomial':
         # analytical formula
-        # logL = np.sum(y*np.log(l) + (1-y)*np.log(1-l))
+        logL = np.sum(y * np.log(y_hat) + (1 - y) * np.log(1 - y_hat))
 
         # but this prevents underflow
-        z = beta0 + np.dot(X, beta)
-        logL = np.sum(y * z - np.log(1 + np.exp(z)))
+        # z = beta0 + np.dot(X, beta)
+        # logL = np.sum(y * z - np.log(1 + np.exp(z)))
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: tests are passing even without this ...

**Comment by jasmainak** [comment]
ready for merge from my end!

File: pyglmnet/pyglmnet.py [hunk 5]
```diff
@@ -132,55 +138,31 @@ def _grad_L2loss(distr, alpha, Tau, reg_lambda, X, y, eta, beta):
...
-            np.exp(eta)
-
-        grad_beta = np.zeros([X.shape[1], ])
-        selector = np.where(z.ravel() <= eta)[0]
-        grad_beta += np.transpose(np.dot((q[selector] - y[selector]).T,
-                                         X[selector, :]))
-        selector = np.where(z.ravel() > eta)[0]
-        grad_beta += np.exp(eta) * \
-            np.transpose(np.dot((1 - y[selector] / q[selector]).T,
-                                X[selector, :]))
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: See how nice it becomes after implementing chain rule :)

**Comment by jasmainak** [comment]
ping @pavanramkumar . Did you get a chance to look at the diff?

File: pyglmnet/utils.py [hunk 6]
```diff
@@ -83,28 +56,8 @@ def label_binarizer(y):
...
-        return np.sum(nu * (-y / np.log1p(np.exp(yhat)) -
-                      np.log(np.log1p(np.exp(yhat)))))
-    elif distr == 'multinomial':
-        y = label_binarizer(y)
-        # yhat is the probability of each output
-        if yhat.ndim != y.ndim:
-            msg = 'yhat and ynull must be (n_samples, n_class) ndarrays'
-            raise ValueError(msg)
-        return np.sum(y * np.log(yhat))
+    from .pyglmnet import _logL
```
Comment (pavanramkumar) [inline pyglmnet/utils.py]: can we import at the top?

File: pyglmnet/pyglmnet.py [hunk 7]
```diff
@@ -14,52 +15,55 @@ def _lmb(distr, beta0, beta, X, eta):
     return _qu(distr, z, eta)
 
 
-def _qu(distr, z, eta):
-    """The non-linearity."""
-    if distr == 'softplus':
+def _qu(distr, z, eta, return_grad=False):
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: can we call this function `_mu`? i believe the tutorial started with $q()$ but we started calling it $\mu()$ in the cheatsheet. it would be nice to make the variable names consistent with docs.

Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: why not compute the gradient only when `return_grad` is `True`? the code is more succinct in your way, but we can be more efficient if we compute the gradient only when it is requested.

File: pyglmnet/utils.py (hunk 6 above)
Comment (jasmainak) [inline pyglmnet/utils.py]: I think you will hit an issue with circular imports if you try to do that. Actually, it's not super clear to me if you need this public facing function. Is this somehow useful to users or just for us? If it's just for us, we can get rid of it safely

File: pyglmnet/pyglmnet.py (hunk 7 above)
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: good catch! I'll fix it

**Comment by jasmainak** [comment]
@pavanramkumar let me know about log likelihood. The other comment is addressed

File: pyglmnet/utils.py (hunk 6 above)
Comment (pavanramkumar) [inline pyglmnet/utils.py]: ok, didn't catch the circular imports issue. right now, it's not meant to be exposed. however, i've seen several papers report log likelihood for model comparison instead of deviance. can we provide it as a score method? for now, i'm happy to hide it from the user in any way you choose.

File: pyglmnet/pyglmnet.py [hunk 8]
```diff
@@ -15,33 +15,39 @@ def _lmb(distr, beta0, beta, X, eta):
     return _mu(distr, z, eta)
 
 
-def _mu(distr, z, eta, return_grad=False):
+def _mu(distr, z, eta):
     """The non-linearity (inverse link)."""
     if distr in ['softplus', 'gamma']:
         qu = np.log1p(np.exp(z))
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: sorry to nitpick, but can all these variables names be `mu` or `grad_mu` instead of `qu` or `grad_qu`. thanks!

**Comment by jasmainak** [comment]
@pavanramkumar I removed `log_likelihood` for now. If users request it later, it's cheap to add it back. The other comment is addressed

File: pyglmnet/pyglmnet.py (hunk 4 above)
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: can we leave these lines commented in? 

when i was first implementing it, i found some strange scenarios when calculating binomial log likelihood in the naive way. this is related to taking log of sigmoid. if you do the algebra you see `np.log(np.exp(z) / (1 + np.exp(z))` which is `z - np.log(1 + np.exp(z)`. i'm concerned the tests are passing because of a very liberal tolerance.

**Comment by pavanramkumar** [comment]
i'm ok with `log_likelihood` but check out my comment regarding the underflow prevention for binomial log likelihood (binomial cross entropy). it's a special case of the well known log-sum-exp trick, see e.g.: https://hips.seas.harvard.edu/blog/2013/01/09/computing-log-sum-exp/ (hint: interview question!)

File: pyglmnet/pyglmnet.py (hunk 4 above)
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: yep we can leave it in. Actually do you a better reference for this than the blog post? I understand the idea, but it wasn't clear to me how it was derived in this particular case and why you need `z`

Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: @jasmainak try deriving it! there are a couple of `log(exp(z))` terms that become `z`.

**Comment by pavanramkumar** [comment]
I will merge this PR now!

File: pyglmnet/pyglmnet.py (hunk 4 above)
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: @pavanramkumar I tried deriving this but somehow I am getting an extra `-z` term. I'll share my derivation with you privately. In any case, I understand that this simplification works only for the canonical link, expit: is that correct? In that case, we should perhaps have an internal check for the link.

In terms of the structure of the code base, I think we should semantically separate 'link' and 'distr'. Right now `_qu(...)` takes 'distr' as an argument but I think it should take 'link'. Then we can have a dictionary which maps the 'distr' to the canonical link.
